### ISSUE TITLE
Add urls for concepts in the concepticon for referencing purposes

### ISSUE THREAD

**Issue by LinguList** [concepticon/concepticon.github.io#6]
This should be done before the official launch (around end of April): we definitely need IDs to be represented as URLs (see also [here](https://github.com/glottobank/lexibank/wiki/An-interchange-format-for-wordlists/_edit)).

**Comment by xrotwang** [comment]
There are a couple of alternatives for these IDs:

- [sluggified](http://en.wikipedia.org/wiki/Semantic_URL#Slug) english gloss: may help with discovery, but may be prone to getting out of sync with the identified concept set. Also requires some infrastructure to make sure IDs are unique when expanding the concepticon.
- [UUID](http://en.wikipedia.org/wiki/Universally_unique_identifier): opaque, bad mnemonic.
- sequential integer number: Again, requires infrastructure to make sure IDs are unique. Seems like it combines the disadvantages of the options listed above, but would make for shorter IDs.

**Comment by cysouw** [comment]
Do I miss something? For example, the concept "contemptible" has a Concepticon ID "1001394", which has a URL http://www.omegawiki.org/DefinedMeaning:1001394. It might not be RDF, but it is nicely humanly-readable :-)

**Comment by xrotwang** [comment]
So that does put omegawiki in control of the URLs. Is this desirable? E.g. how easy is it to expand the concepticon? Can new entries in omegawiki be created programmatically?

An alternative could be minting concept URLs in a domain like concepticon.clld.org, and considering the omegawiki URLs as just one mapping?

**Comment by cysouw** [comment]
That was the basic idea when we starting the concepticon: putting omegawiki in control! There is nothing better, and it is easy to expand when necessary (it's a wiki after all). We thought for a long time about how to best define concepts, and omegawiki is currently the most extensive free and open collection of definitions available. When somebody has a better idea how to define meaning, then we could easily switch. But be aware that such a switch is a *lot* of work!

**Comment by xrotwang** [comment]
Ok. So the idea is to use omegawiki as the main reference point and as source of identifiers, and to use GitHub pages to publish the mappings from omegawiki to other concept lists? While I agree that both omegawiki and GitHub are probably quite stable and will be around for some time, I think the additional effort of making the mapping data available from a domain under direct control of the project may add a layer of abstraction which would allow migrating from GitHub to another hosting platform.

**Comment by xrotwang** [comment]
To add some context: I'm after an exchange format for wordlists, which should use a concept ID to reference a concept rather than identify it by gloss. Thus, to merge wordlists or aggregate lexical data, I'd need a service that allows looking up concepticon IDs given IDs from other concept lists, e.g. http://wold.clld.org/meaning/3-11

**Comment by cysouw** [comment]
github is immaterial from my perspective: we just currently keep the data there.

There is nothing agains providing a mirror-set of URLs, like concepticon.clld.org/1001394 if you think that one is more stable.

You can also drag all relevant information from omegawiki to populate those URLs, but please note the crucial different between "defined meanings' and "expressions"!!! 

OmegaWiki is not the most active of all internet-projects, but they keep working on it, and they do have a lot of data.

Your example of WOLD ist nice, because it has a definition and an example. However, only a small subset of WOLD has such information. OmegaWiki currently simply has much more background information. That is why we decided to start there

**Comment by xrotwang** [comment]
So using the concepticon data to build a service allowing queries like this:
```python
$ curl http://concepticon.clld.org/concept/1001394.json
{
    "description": "...", 
    "id": "http://omegawiki.org/DefinedMeaning:1001394",
    "sameAs": {
        "http://concepticon.clld.org/conceptlist/swadesh-100": "..."
    } 
}
```
would be in line with your plans for the project?

**Comment by cysouw** [comment]
yep, that is basically the kind of information we are collecting, see https://github.com/concepticon/lists/tree/master/conceptlists. The list from Swadesh in 1955 has local IDs, the GLOSSes as used by Swadesh (we try to be a precise as possible here, typing off the way how things are written in the original source!), and a link to an OmegaWiki ID, e.g. https://github.com/concepticon/lists/blob/master/conceptlists/Swadesh-1955-100.tsv

For practical reasons, we are working from the original conceptlists, and linking them to OmegaWiki.

**Comment by LinguList** [comment]
When working on getting rid of those cases where we left stuff unlinked (our NANs in the linked lists), I figured that on the long run we may run into some practicality issues with OmegaWiki, since we should aim for clearing all NANs and having everything linked completely. Here, we have some simple cases which are unproblematic, like those cases where we can easily find a concept in OWIKI, like, e.g. "frost", which occurs in a Chinese source. We have, however, also cases where it is just getting too specific, like, e.g. in the case of classifiers in Chinese but also in other languages with classifiers: since classifiers are basically not there in OWIKI, we need to add all of these as specific definitions. This is of course possible, but it is not really feasible in terms of time managment, since it costs a lot of time, and we cannot really entrust this work to students (unless they have a good knowledge of the respective languages). So the question here is whether we can find some intermediate solution which would allow us to identify these things on a working basis, without breaking them down to NANs, and later map them to OWIKI by creating new entries in one flush? So maybe, in the end, it could be a good idea to create our unique internal identifiers (maybe at least for the NAN-cases), before we clear them by adding them to OWIKI?

The most annoying thing of OWIKI, though @cysouw is right that it seems to be the best that is there on the web, is that it is really slow to work with the interface. So adding entries turns out to consume a lot of time, and it would be much more efficient to do this for multiple entries at once, rather than doing it on a per-source basis.

**Comment by cysouw** [comment]
We could of course use @xrotwang 's proposal, and rebuild our own identifiers, using the information from OmegaWiki as a basis (and refer to it). Basically, use http://concepticon.clld.org/concept/1001394 with a copy of the relevant OmegaWiki information on our own server.

When we want to add new concepts, either use OmegaWiki as a basis (when there is already information there), or simply add new concepts locally at clid.org without reference to OmegaWiki.

What I would like to retain from OmegaWiki is the necessity to have an explicit definition for each concept, i.e. a sentence (or longer) trying to define what the concept is about.

We might then also replace the number with our GLOSS, i.e.  http://concepticon.clld.org/concept/contemptible/, leading to something as shown below. What I would very much like to have though, is an easy way to manually edit this information in text files, roughly like we are currently using.

$ curl http://concepticon.clld.org/concept/contemptible.json
{
    "definedMeaning": {
        {"langoid": "Castilian",
         "definition": "Que merece desprecio."},
        {"langoid": "Dutch",
         "definition": "Verachting verdienend."},
         ..., 
    "sameAs": {
        "http://omegawiki.org/DefinedMeaning:1001394",
        "http://concepticon.clld.org/conceptlist/swadesh-100": "..."
    } 
}

**Comment by LinguList** [comment]
I think this would make it much more practicable at the moment, and we
could still try to go for explicit mapping with OWID. It has the further
advantage that we are not necessarily bound to the definitions in OWIKI,
since my experience here in Paris was that some of them would confuse
the scholars here more then helping to explain. We can, of course, never
reach a full consensus here, but we could offer the possibility to
revise and refine definitions (requiring a PR etc.) in a much easier way
then by forcing people to change OWIKI directly.

The question now is, in terms of time managment (we should had for a
beta-release during the end of April), what we need for this new
infrastructure, and how much time it will require to come up with it.
Are there any ideas or suggestions?

**Comment by xrotwang** [comment]
ok, here's my proposal

Concepticon in CLLD
===================

Since the data and mappings provided by the concepticon project are an important
building block to interoperability between lexical resources, they should be
published as clld app (edited by Cysouw and List), thereby providing an API
which can be used by tools and methods.


Data model
----------

The main entities of the data model are

- ConcepticonConcept: An "authoritative" or "hub" concept, which can be used
  to relate other concepts to or directly by referencing it from wordlists.
  A ConcepticonConcept is identified by its unique English gloss. All
  ConcepticonConcepts have a description or example, possibly imported from
  OmegaWiki.
- ConceptList: A list of concepts typically created to collect wordlists;
  e.g. the swadesh list, but maybe also OmegaWiki's DefinedMeanings.
- SourceConcept: A concept as it appears in a ConceptList.
- Relation (better term?): Relates SourceConcepts to ConcepticonConcepts,
  specifying a relation from "sameAs|narrower|broader".
  Each SourceConcept is assumed to be related to at least one ConcepticonConcept
  via a sameAs relation.

clld app
--------

With this data model, a clld app can provide:

- URLs identifying SourceConcepts and ConcepticonConcepts (as well as ConceptLists).
- an API allowing discovery of related concepts given either a Source- or
  ConcepticonConcept.


Data curation
-------------

Data curation could proceed pretty much as done now, i.e. the data can be
curated as a set of csv files maintained in a GitHub repository. The clld app
can import well-defined releases of this repository. The set of files may
look as follows:

- One csv file listing ConcepticonConcepts:
        ID      Gloss   Description
- One csv file per ConceptList, e.g. swadesh-1955.csv:
        ID      Gloss   ConcepticonConceptID
- One csv file per ConceptList specifying additional relations:
        ID      ConcepticonConceptID    relation

The data repository could be equipped with a small python package providing data
integrity checks which can be run after each push as Travis-CI jobs.

**Comment by LinguList** [comment]
I really like this proposal, especially also due to its practical implications: As our work in the past has shown, taking OWIKI as the main reference of concept mappings bears certain shortcomings that I consider now as too grave to make OWIKI our standard reference and seem to cry for an alternative authoritative solution (in which the mappings to OWIKI will still play a role, but not as a necessary condition). Just to summarize my experience from the past:

* our linking was implicitly always a "sameAs" linking, but we will loose important information if we restrict ourself to it (compare the two readings of "child" as "descendant" and "young human", which will be treated separately, although most scholars would just have the broader concept in their wordlist).
* maintaining the data is a pain in the neck at the moment, since it requires to manually update the concepticon.tsv file, from which many errors in the current version result (we have multiple base glosses linking to different OWIKI-IDs, and it is hard to get track of all cases).
* adding new data is complicated, since the workflow requires not only the mapping to existing ones, but for each new concept not linked to OWIKI so far, we need to go to OWIKI and either search for it or add it, and this is really slow and even doesn't guarantee that we are consistent here, since OWIKI has many concepts which are virtually identical but are linked to different IDs. In the past, we had our workaround here by assigning NANs to concepts which we were too lazy to map, but this turned out to lead to a great loss in accuracy, with, for example, many classifiers in Sino-Tibetan languages not being linked so far.
* representing comparisons between two or more concept lists is lossy in so far as valuable similarities between concepts which are not retrievable via samaAS links are not taken care of, although they play a crucial role in historical linguistic practice.
* modifying mis-mappings of our concepts to the concepticon is very complicated, although we have quite a lot of mismappings to the concepticon. Some of these I only detected recently, when I used the OWIKI definitions to create the concept list for the Sino-Tibetan database: the linguists at CRLAO Paris would tell me that they find most of the definitions not helpful at all, since they would often not be in line with common "intuitive" definitions, especially of the basic vocabulary concepts. 
* our concept glosses are rather ad-hoc at the moment, and I had to re-phrase them for the Sino-Tibetan database, since they would rather confuse the collaborators, especially since we don't distinguish between verbs and nouns (WOLD style, with "the " and "to " preceding is now used, and my collaborators here are much more content with this)

These are only the points from the practical side which I experienced in the past, and I think they should not be underestimated, since this is a concrete use-case which could become some model for future data-collections, and we don't want to scare off people here.

We had good reasons to assume to proceed as we proceeded in the past, but now the concrete working experience has shown to me that we need try to improve.

What follows from @xrotwang's proposal, however, is in concrete that we won't be able to head for a release around end of April, since much more work needs to be done in order to meet the outline. This involves a re-inspection of all glosses we collected so far (re-formulating things like "DRINK ACTION" vs. "DRINK FLUID" to some clearer concept gloss, like "TO DRINK" vs. "THE DRINK"). This will take time, especially since it cannot be entrusted to student collaborators to do this job (one needs some good intuition especially regarding the different basic vocabulary lists, in order to do this properly, and I doubt that we will find student collabs capable of doing this). Another thing to be done will involve re-evaluating the links which have been drawn between concept lists, and I would much like to have some server-tool for this task, since from my experience in the past, it is very tedious to have this done with the tools that we created so far, especially since mapping one list often also involves re-assessing other lists, which is not feasible to be done on a pure csv-basis alone (however, for this, the EDICTOR can be used in a modified fashion, and since most of the mappings are work on my table, I'll try to look into it so that I find a good way to maintain consistency here, and it is basically independent of the concepticon as a whole).

Two burning questions of which I think we should decide them quickly are:

1. How do we add new concepts, and whose authority decides whether they should be added or not? If we give potential external users too much freedom here, it is quite likely that it will get very messy, since people may overlook concepts which are already in the database and propose new ones (this happens in all projects that tried similar things so far, including STEDT, REFLEX, and also OWIKI).
2. How do we re-adjust our csv format for mappings to the concepticon? Currently, we put one relation in one line and reduplicate the row, but if we add three different types of relations, I can't quite think of how this would best look in the csv file.

**Comment by xrotwang** [comment]
As far as I can tell, the only problem which prevents us from starting with the data as it is now, is the idea/requirement to use the gloss of a ConcepticonConcept as its ID, e.g. having URLs like  http://concepticon.clld.org/concept/contemptible/
If we don't want to create lots of URLs we might not be happy with in the future, this would require re-examining all glosses as described by @LinguList.

If on the other hand we drop this requirement and switch to opaque identifiers, like UUIDs or the like, this problem would disappear, right?

**Comment by LinguList** [comment]
Yes! Re-glossing would then just be something that we could do on a step by step basis. And it would also have the further advantage that we do not need to live for ever with a gloss which we once defined but afterwards are no longer happy with (this is very likely to happen, since experience from working on our glosses in the actual concepticon also revealed that we would change them from time to time, as we acquired more data and just realized that the gloss was misleading, ambiguous, etc.).
